### PR TITLE
New structure for the random, updated test files, other minor changes.

### PR DESCRIPTION
**Pull request by Kubasinska**
1. The structure of the random has changed. Both the random module and the numpy.random module are now used within the model. The two generators are created when you instantiate the model by passing a seed to the init (by default no seed is passed and the model is then random).
2. Created a file for TravisCI, now travis correctly tests the pyhton code through unittest. Next step is to replace unittest with pytest so that we have more control in the travis console.
3. Additional tests added
4. Changes to the extra module, now the functions of this module can also take an instance of the random generator as argument. 
5. Minor changes in syntax

### CODE REVIEW

File: mesa/extra.py [hunk 1]
```diff
@@ -33,8 +30,7 @@ def find_neighb(netname, togo, found, border):
 def wedding_proximity_with(ego, pool): # returns a list of proximities with ego. Careful not to shuffle it!
     l = np.array([
         (social_proximity(ego,x) + 
-         (4 - abs(x.hobby - ego.hobby)) / 4 ) / 2 for x in pool
-        ])
+         (4 - abs(x.hobby - ego.hobby)) / 4 ) / 2 for x in pool])
```
Comment (mariopaolucci) [inline mesa/extra.py:33]: I disagree. Why?

File: mesa/extra.py [hunk 2]
```diff
@@ -51,21 +47,21 @@ def social_proximity(ego:Person, alter:Person):
                  ] else 0
     return acc
 
-def at_most(n, a):
-    return a if len(a) < n else rng.choice(a,n)
+def at_most(n, a, rng_istance):
+    return a if len(a) < n else rng_istance.choice(a,n)
 
 # must keep the random generator here? Or in the main function? In which case.. mmh..
```
Comment (mariopaolucci) [inline mesa/extra.py:53]: you made the choice so the comment must go.

File: mesa/extra.py [hunk 3]
```diff
@@ -51,21 +47,21 @@ def social_proximity(ego:Person, alter:Person):
...
     sump = sum(p)
     #minp = min(p)
     #maxp = max(p)
     p = [i/sump for i in p]
-    return  rng.choice(agentset, n, replace = False, p=p)
+    return  rng_istance.choice(agentset, n, replace=False, p=p)
 
 def weighted_one_of(agentset, weight_function):
-    return random.default_rng().weighted_n_of(1, agentset, weight_function) #Wrong declaration
+    return weighted_n_of(1, agentset, weight_function)
```
Comment (mariopaolucci) [inline mesa/extra.py:64]: This needs the rng_instance too, after the change above.

File: mesa/mesaPROTON_OC.py [hunk 4]
```diff
@@ -82,9 +87,9 @@ def __init__(self):
         self.num_oc_families = 8
         self.education_modifier = 1.0
 
-        # todo: Adding a statement for city selection
         # Folders definition
-        self.cwd = os.getcwd()
+        self.mesa_dir = os.getcwd()
+        self.cwd = self.mesa_dir[:-5]
```
Comment (mariopaolucci) [inline mesa/mesaPROTON_OC.py:92]: this sounds like magic. Is it os-independent? Why -5? Don't they have a name?

File: mesa/mesaPROTON_OC.py [hunk 5]
```diff
@@ -246,11 +249,8 @@ def soc_add_more_friends(self, targets):
         for x in targets:
             support_set = limited.extraction(schedule.agents.remove(x))
             if support_set:
-                x.makeFriends(random.choices(support_set,
-                                             weights=[
-                                                 self.criminal_tendency_subtractfromme_for_inverse_weighted_extraction - y.criminal_tendency
-                                                 for y in support_set],
-                                             k=1))
+                x.makeFriends(self.rng.choice(support_set, size=1,
```
Comment (mariopaolucci) [inline mesa/mesaPROTON_OC.py:252]: Line too long.

File: mesa/mesaPROTON_OC.py [hunk 6]
```diff
@@ -449,21 +448,21 @@ def conclude_wedding(ego, partner):
 staticmethod(conclude_wedding)
 
 
-if __name__ == "__main__":
```
Comment (mariopaolucci) [inline mesa/mesaPROTON_OC.py:452]: Better proposing to remove it altogether.

File: mesa/testProton.py [hunk 7]
```diff
@@ -37,7 +38,8 @@ def test_networks(self):
         for i in range(0,10): 
             pp.Person(m)
         for i in range(0,10):
-            for l in links[i]: pp.Person.persons[i].neighbors.get('friendship').add(pp.Person.persons[l])
```
Comment (mariopaolucci) [inline mesa/testProton.py:40]: In general, it's better if you don't make this kind of formatting change (except maybe on a dedicated commit)

File: mesa/Person.py [hunk 8]
```diff
@@ -65,15 +64,15 @@ def age(self):
         
     def random_init(self):
         self.randomfriends()
-        self.education_level = random.choice(range(0,4))
+        self.education_level = self.m.rng.choice(range(0,4))
```
[the remark this answers is not in the document]
Comment (Kubasinska) [reply mesa/Person.py:67]: I think this form is more explicit, the generator that is used is that of the model.

File: mesa/extra.py (hunk 1 above)
Comment (Kubasinska) [reply mesa/extra.py:33]: That's my mistake.

File: mesa/extra.py (hunk 3 above)
Comment (Kubasinska) [reply mesa/extra.py:64]: my mistake again!

File: mesa/mesaPROTON_OC.py (hunk 4 above)
Comment (Kubasinska) [reply mesa/mesaPROTON_OC.py:92]: It's true, that seems unclear.

File: mesa/extra.py [hunk 9]
```diff
@@ -51,21 +48,21 @@ def social_proximity(ego:Person, alter:Person):
...
-    return  rng.choice(agentset, n, replace = False, p=p)
+    return  rng_istance.choice(agentset, n, replace=False, p=p)
 
-def weighted_one_of(agentset, weight_function):
-    return random.default_rng().weighted_n_of(1, agentset, weight_function) #Wrong declaration
+def weighted_one_of(agentset, weight_function, rng_istance):
+    return weighted_n_of(1, agentset, weight_function, rng_istance)
 
 def pick_from_pair_list(a_list_of_pairs):
     return weighted_one_of(a_list_of_pairs, lambda x: x[-1])[0]
```
Comment (mariopaolucci) [inline mesa/extra.py:68]: qui manca il generatore

Comment (Kubasinska) [reply mesa/extra.py:68]: my mistake again. I fixed

**Pull request merged**